File: modules/module-interop/android/src/main/java/com/moduleinterop/ModuleInterop.kt
package com.moduleinterop

import android.content.Context
import expo.modules.logger.ExpoLoggerCore
import expo.modules.storage.StorageCore

/**
 * Centralized Interop Layer for all module-to-module communication
 * 
 * This class provides a single entry point for cross-module calls,
 * eliminating the need for reflection or direct module dependencies.
 * 
 * IMPORTANT: This is a STATELESS FACADE - it delegates to Core classes.
 * Each Core class owns its own state and business logic.
 * 
 * In production, this will live inside cg-webview and provide access
 * to AppsFlyer, Firebase, Snowplow, and other modules.
 */
class ModuleInterop private constructor(private val context: Context) {
    
    // ========================================
    // LOGGER INTEROP (delegates to ExpoLoggerCore)
    // ========================================
    
    fun logInfo(message: String) {
        ExpoLoggerCore.getInstance().logInfo(message)
    }
    
    fun getLogCount(): Int {
        return ExpoLoggerCore.getInstance().getLogCount()
    }
    
    fun resetCount() {
        ExpoLoggerCore.getInstance().resetCount()
    }
    
    // ========================================
    // STORAGE INTEROP (delegates to StorageCore)
    // ========================================
    
    fun setItem(key: String, value: String) {
        StorageCore.getInstance().setItem(key, value)
    }
    
    fun getItem(key: String): String? {
        return StorageCore.getInstance().getItem(key)
    }
    
    // ========================================
    // SINGLETON
    // ========================================
    
    companion object {
        @Volatile
        private var instance: ModuleInterop? = null
        
        @JvmStatic
        fun getInstance(context: Context): ModuleInterop {
            return instance ?: synchronized(this) {
                instance ?: ModuleInterop(context).also { 
                    instance = it
                }
            }
        }
    }
}

File: ios/NewArchSpike/TurboModules/ModuleInterop.swift
import Foundation
import ExpoLogger
import ExpoStorage

/**
 * Centralized Interop Layer for all module-to-module communication
 *
 * This class provides a single entry point for cross-module calls,
 * eliminating the need for reflection or direct module dependencies.
 *
 * In production, this will live inside cg-webview and provide access
 * to AppsFlyer, Firebase, Snowplow, and other modules.
 */
@objc class ModuleInterop: NSObject {
    @objc static let shared = ModuleInterop()
    
    private override init() {
        super.init()
        // Private initializer to enforce singleton pattern
    }
    
    // ========================================
    // LOGGER INTEROP (delegates to ExpoLoggerCore)
    // ========================================
    
    @objc func logInfo(_ message: String) {
        ExpoLoggerCore.shared.logInfo(message)
    }
    
    @objc func getLogCount() -> Int {
        return ExpoLoggerCore.shared.getLogCount()
    }
    
    @objc func resetCount() {
        ExpoLoggerCore.shared.resetCount()
    }
    
    // ========================================
    // STORAGE INTEROP (delegates to StorageCore)
    // ========================================
    
    @objc func setItem(_ key: String, value: String) {
        StorageCore.shared.setItem(key, value: value)
    }
    
    @objc func getItem(_ key: String) -> String? {
        return StorageCore.shared.getItem(key)
    }
}

